src/timestamp_generator.cpp

```diff
@@ -52,24 +61,31 @@ int64_t MonotonicTimestampGenerator::next() {
   }
 }
 
+// This is guaranteed to return a monotonic timestamp. If clock skew is deteced
+// then this method will increment the last timestamp.
 int64_t MonotonicTimestampGenerator::compute_next(int64_t last) {
-  int64_t millis = last / 1000;
-  int64_t counter = last % 1000;
-
-  int64_t now = get_time_since_epoch_ms();
-
-  if (millis >= now) {
-    if (counter == 999) {
-      LOG_WARN("Sub-millisecond counter overflowed, some query timestamps will not be distinct");
-    } else {
-      ++counter;
+  int64_t current = get_time_since_epoch_us();
+
+  if (last >= current) { // There's clock skew
+    // If we execeed our warning threshhold then warn periodically that clock
+    // skew has been detected.
+    if (warning_threshold_us_ >= 0 && last > current + warning_threshold_us_) {
+      int64_t now = get_time_since_epoch_ms();
+      int64_t last_warning = last_warning_.load();
+      if (now > last_warning + warning_interval_ms_ &&
```

Comment:
A real fix is to use a monotonic system clock to manage the warning interval. The problem is getting that to work properly on all supported platforms (esp. in Java and Python). It might be overkill for handling the warnings, but I'm going at least investigate how hard it would be to implement a millisecond precision monotonic clock for the C++ driver's major platforms (BSD, Linux, Windows).